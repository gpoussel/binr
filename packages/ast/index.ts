export * from "./src/evaluation";
export * from "./src/nodes";
export * from "./src/visitor";
