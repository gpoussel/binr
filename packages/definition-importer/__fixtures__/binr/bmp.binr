#name "BMP Format"
#extension ".bmp"

struct FileHeader {
    uint:16    type; 
    uint:32    size; 
    uint:16    reserved[2];
    uint:32    offBits;
}

struct InfoHeader {
    uint:32 size;
    uint:32 width;
    uint:32 height;
    uint:16 planes;
    uint:16 bitCount;
    uint:32 compression;
    uint:32 sizeImage;
    uint:32 xPelsPerMeter;
    uint:32 yPelsPerMeter;
    uint:32 clrUsed;
    uint:32 clrImportant;
}

struct BmpFile {
    FileHeader fileHeader;
    InfoHeader infoHeader;
}