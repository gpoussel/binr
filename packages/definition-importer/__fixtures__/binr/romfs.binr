#name "ROMFS Format"
#extension ".romfs"
#endian "big"

// Source: https://www.sweetscape.com/010editor/repository/files/ROMFS.bt

struct RomFSHeader {
    char magicBytes[8];
    uint:32 size;
    int:32 checksum;
    cstring volumeName;
}

enum RomFSEntryType extends uint:3 {
    HARD_LINK = 0,
    DIRECTORY = 1,
    FILE = 2,
    SYMLINK = 3,
    BLOCK_DEV = 4,
    CHAR_DEV = 5,
    SOCKET = 6,
    FIFO = 7
}

struct RomFSObjectHdr {
    // int:28 nextEntryPosition;
    // int:1 executable;
    // RomFSEntryType type;
}

struct RomFSObject {
    RomFSObjectHdr header;
}

struct RomfsFile {
    RomFSHeader header;
    RomFSObject entry;
}
