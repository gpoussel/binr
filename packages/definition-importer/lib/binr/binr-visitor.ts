import {
  Annotation,
  ArraySelector,
  ArrayValue,
  ArrayValueElement,
  BinaryExpression,
  BitmaskDeclarationElement,
  BitmaskDeclarationStatement,
  BlockStatement,
  BooleanValue,
  CaseSwitchElement,
  Definition,
  EnumDeclarationElement,
  EnumDeclarationStatement,
  Expression,
  ExpressionArraySelector,
  ExpressionArrayValueElement,
  IdentifierValue,
  IfElseStatement,
  IfStatement,
  NamedType,
  NumberValue,
  Operator,
  PostfixExpression,
  PrefixExpression,
  RestrictedType,
  Statement,
  StringValue,
  StructDeclarationStatement,
  SwitchStatement,
  TernaryExpression,
  Type,
  UndefinedArrayValueElement,
  UntilExpressionArraySelector,
  Value,
  ValueSwitchLabel,
  VariableDeclaration,
  VariableDeclarationStatement,
} from "@binr/ast";
import { CstParser } from "chevrotain";
import { concat, each, flatMap, get, has, keys, map, size, times } from "lodash";

const OPERATOR_MAPPING: { [key: string]: Operator } = {
  BooleanOrToken: Operator.BOOLEAN_OR,
  BooleanAndToken: Operator.BOOLEAN_AND,
  BinaryOrToken: Operator.BINARY_OR,
  BinaryXorToken: Operator.BINARY_XOR,
  BinaryAndToken: Operator.BINARY_AND,
  DoubleEqualsToken: Operator.DOUBLE_EQUALS,
  DifferentToken: Operator.DIFFERENT,
  ShiftRightToken: Operator.SHIFT_RIGHT,
  ShiftLeftToken: Operator.SHIFT_LEFT,
  UnsignedShiftRightToken: Operator.UNSIGNED_SHIFT_RIGHT,
  GreaterToken: Operator.GREATER,
  LessToken: Operator.LESS,
  MultiplicationToken: Operator.MULTIPLICATION,
  DivisionToken: Operator.DIVISION,
  PlusToken: Operator.PLUS,
  MinusToken: Operator.MINUS,
  ModuloToken: Operator.MODULO,
};

export function getVisitor(parser: CstParser) {
  class Visitor extends parser.getBaseCstVisitorConstructorWithDefaults() {
    constructor() {
      super();
      this.validateVisitor();
    }

    public definition(ctx: any): Definition {
      const statements = this.visitAll(ctx, "topLevelClause");
      return new Definition(statements);
    }

    public topLevelClause(ctx: any): Statement {
      const annotations = this.visitAll(ctx, "annotationClause");
      if (has(ctx, "structureClause")) {
        const structureCtx = ctx.structureClause[0].children;
        // TODO "exported" flag (ExportToken)
        const name = this.getIdentifierName(structureCtx.IdentifierToken[0]);
        const statements: Statement[] = this.visitAll(structureCtx, "statementClause");
        return new StructDeclarationStatement(name, [], new BlockStatement(statements), annotations);
      }
      if (has(ctx, "enumClause")) {
        const enumCtx = ctx.enumClause[0].children;
        const name = this.getIdentifierName(enumCtx.IdentifierToken[0]);
        const entries = times(enumCtx.IdentifierToken.length - 1, (i) => {
          const key = this.getIdentifierName(enumCtx.IdentifierToken[i + 1]);
          const value = this.visit(enumCtx.numberClause[i]);
          return new EnumDeclarationElement(key, value);
        });
        const parentType = this.visit(enumCtx.typeReferenceClause);
        return new EnumDeclarationStatement(parentType, name, entries, annotations);
      }
      if (has(ctx, "bitmaskClause")) {
        const bitmaskCtx = ctx.bitmaskClause[0].children;
        const name = this.getIdentifierName(bitmaskCtx.IdentifierToken[0]);
        const entries = times(bitmaskCtx.IdentifierToken.length - 1, (i) => {
          const key = this.getIdentifierName(bitmaskCtx.IdentifierToken[i + 1]);
          const value = this.visit(bitmaskCtx.numberClause[i]);
          return new BitmaskDeclarationElement(key, value);
        });
        const parentType = this.visit(bitmaskCtx.typeReferenceClause);
        return new BitmaskDeclarationStatement(parentType, name, entries, annotations);
      }
      throw new Error();
    }

    public statementClause(ctx: any): Statement {
      return this.visitFirst(ctx, "fieldClause", "IfStatement", "BlockStatement", "SwitchStatement");
    }

    public IfStatement(ctx: any): IfElseStatement | IfStatement {
      const condition = this.visit(ctx.Expression[0]);
      const trueStatement = this.visit(ctx.statementClause[0]);
      if (size(ctx.statementClause) > 1) {
        const falseStatement = this.visit(get(ctx.statementClause, 1));
        return new IfElseStatement(condition, trueStatement, falseStatement);
      }
      return new IfStatement(condition, trueStatement);
    }

    public fieldClause(ctx: any): VariableDeclarationStatement {
      const type = this.visit(ctx.typeReferenceClause);
      const name = this.getIdentifierName(get(ctx.IdentifierToken, 0));
      const annotations: Annotation[] = this.visitAll(ctx, "annotationClause");
      let arraySelector: ArraySelector | undefined = undefined;
      if (has(ctx, "BoxMemberUntilExpression")) {
        arraySelector = this.visit(ctx.BoxMemberUntilExpression[0]);
      } else if (has(ctx, "BoxMemberExpression")) {
        arraySelector = this.visit(ctx.BoxMemberExpression[0]);
      }
      const declaration: VariableDeclaration = new VariableDeclaration(
        name,
        undefined,
        arraySelector,
        [],
        undefined,
        [],
      );
      return new VariableDeclarationStatement(type, [], undefined, [declaration], annotations);
    }

    public BlockStatement(ctx: any): BlockStatement {
      return new BlockStatement(this.visitAll(ctx, "statementClause"));
    }

    public SwitchStatement(ctx: any): SwitchStatement {
      const testExpression = this.visit(ctx.Expression[0]);
      const switchElements = this.visitAll(ctx, "switchInnerClause");
      return new SwitchStatement(testExpression, switchElements);
    }

    public switchInnerClause(ctx: any): CaseSwitchElement {
      const value = this.visit(ctx.valueClause[0]);
      const statement = this.visit(ctx.BlockStatement[0]);
      return new CaseSwitchElement([new ValueSwitchLabel(value)], [statement]);
    }

    public Expression(ctx: any): Expression {
      return this.visit(ctx.AssignmentExpression);
    }

    public AssignmentExpression(ctx: any): Expression {
      if (has(ctx, "QuestionToken")) {
        const condition = this.visit(ctx.BinaryExpression);
        const trueExpression = this.visit(ctx.AssignmentExpression[0]);
        const falseExpression = this.visit(ctx.AssignmentExpression[1]);
        return new TernaryExpression(condition, trueExpression, falseExpression);
      }
      return this.visit(ctx.BinaryExpression);
    }

    public BinaryExpression(ctx: any): Expression {
      let currentExpression = this.visit(ctx.UnaryExpression[0]);
      each(ctx.ExpressionToken, (token, i) => {
        const operator = this.visit(token);
        const childExpression = this.visit(ctx.UnaryExpression[i + 1]);
        currentExpression = new BinaryExpression(currentExpression, childExpression, operator);
      });
      return currentExpression;
    }

    public UnaryExpression(ctx: any): Expression {
      if (has(ctx, "DoublePlusToken")) {
        return new PrefixExpression(this.visit(ctx.UnaryExpression), Operator.DOUBLE_PLUS);
      }
      if (has(ctx, "DoubleMinusToken")) {
        return new PrefixExpression(this.visit(ctx.UnaryExpression), Operator.DOUBLE_MINUS);
      }
      if (has(ctx, "PlusToken")) {
        return new PrefixExpression(this.visit(ctx.UnaryExpression), Operator.PLUS);
      }
      if (has(ctx, "MinusToken")) {
        return new PrefixExpression(this.visit(ctx.UnaryExpression), Operator.MINUS);
      }
      if (has(ctx, "TildaToken")) {
        return new PrefixExpression(this.visit(ctx.UnaryExpression), Operator.TILDA);
      }
      if (has(ctx, "ExclamationToken")) {
        return new PrefixExpression(this.visit(ctx.UnaryExpression), Operator.EXCLAMATION);
      }
      return this.visit(ctx.PostfixExpression);
    }

    public PostfixExpression(ctx: any): Expression {
      const innerExpression = this.visit(ctx.MemberCallNewExpression);
      if (has(ctx, "DoublePlusToken")) {
        return new PostfixExpression(innerExpression, Operator.DOUBLE_PLUS);
      }
      if (has(ctx, "DoubleMinusToken")) {
        return new PostfixExpression(innerExpression, Operator.DOUBLE_MINUS);
      }
      return innerExpression;
    }

    public MemberCallNewExpression(ctx: any): Expression {
      const innerExpression = this.visit(ctx.PrimaryExpression);
      // TODO MemberCallNewExpressionExtension
      return innerExpression;
    }

    public BoxMemberExpression(ctx: any): ExpressionArraySelector {
      return new ExpressionArraySelector(this.visit(ctx.Expression));
    }
    public BoxMemberUntilExpression(ctx: any): UntilExpressionArraySelector {
      return new UntilExpressionArraySelector(this.visit(ctx.Expression));
    }

    public PrimaryExpression(ctx: any): Expression {
      if (has(ctx, "numberClause")) {
        return this.visit(ctx.numberClause[0]);
      }
      if (has(ctx, "IdentifierToken")) {
        return new IdentifierValue(this.getIdentifierName(ctx.IdentifierToken[0]));
      }
      if (has(ctx, "ParenthesisExpression")) {
        return this.visit(ctx.ParenthesisExpression);
      }
      if (has(ctx, "ArrayLiteral")) {
        return this.visit(ctx.ArrayLiteral);
      }
      if (has(ctx, "StringLiteralToken")) {
        return new StringValue(ctx.StringLiteralToken[0].image);
      }
      throw new Error();
    }

    public ParenthesisExpression(ctx: any): Expression {
      return this.visit(ctx.Expression);
    }

    public ArrayLiteral(ctx: any): ArrayValue {
      return new ArrayValue(flatMap(ctx.ArrayLiteralContent, this.visit.bind(this)));
    }

    public ArrayLiteralContent(ctx: any): ArrayValueElement[] {
      if (has(ctx, "ElementList")) {
        return this.visit(ctx.ElementList);
      }
      if (has(ctx, "Elision")) {
        return this.visit(ctx.Elision);
      }
      throw new Error();
    }

    public ElementList(ctx: any): ExpressionArrayValueElement[] {
      return concat(
        [new ExpressionArrayValueElement(this.visit(ctx.AssignmentExpression))],
        map(ctx.ElementListEntry, this.visit.bind(this)),
      );
    }

    public ElementListEntry(ctx: any): ExpressionArrayValueElement {
      return new ExpressionArrayValueElement(this.visit(ctx.AssignmentExpression));
    }

    public Elision(ctx: any): UndefinedArrayValueElement[] {
      return times(size(ctx.CommaToken), () => new UndefinedArrayValueElement());
    }

    public ExpressionToken(ctx: any): Operator {
      const tokenName = keys(ctx)[0];
      return OPERATOR_MAPPING[tokenName];
    }

    public valueClause(ctx: any): Value {
      if (has(ctx, "StringLiteralToken")) {
        return new StringValue(JSON.parse(ctx.StringLiteralToken[0].image));
      }
      if (has(ctx, "TrueToken")) {
        return new BooleanValue(true);
      }
      if (has(ctx, "FalseToken")) {
        return new BooleanValue(false);
      }
      // That's a number
      return new NumberValue(this.visit(ctx.numberClause[0]));
    }

    public numberClause(ctx: any): number {
      if (has(ctx, "NumberDecimalLiteralToken")) {
        return parseInt(ctx.NumberDecimalLiteralToken[0].image, 10);
      }
      if (has(ctx, "NumberHexadecimalLiteralToken")) {
        return parseInt(ctx.NumberHexadecimalLiteralToken[0].image.substring(2), 16);
      }
      if (has(ctx, "NumberBinaryLiteralToken")) {
        return parseInt(ctx.NumberBinaryLiteralToken[0].image.substring(2), 2);
      }
      throw new Error();
    }

    public typeReferenceClause(ctx: any): Type {
      const type = this.getIdentifierName(get(ctx.IdentifierToken, 0));
      const baseType = new NamedType(type, [], false);
      if (has(ctx, "ColonToken")) {
        const typeRestriction = this.visit(ctx.numberClause[0]);
        return new RestrictedType(baseType, typeRestriction);
      }
      return baseType;
    }

    public annotationClause(ctx: any): Annotation {
      const name = this.getIdentifierName(ctx.IdentifierToken[0]);
      const value = this.visit(ctx.valueClause[0]);
      return new Annotation(name, value);
    }

    private visitAll(ctx: any, propertyName: string): any[] {
      return map(get(ctx, propertyName), this.visit.bind(this));
    }

    private visitFirst(ctx: any, ...propertyNames: string[]): any {
      for (let i = 0; i < propertyNames.length; ++i) {
        if (has(ctx, propertyNames[i])) {
          return this.visit(get(ctx, propertyNames[i]));
        }
      }
      throw new Error(
        `Context does not contain expected property name (expected: ${propertyNames}, was: ${keys(ctx)})`,
      );
    }

    private getIdentifierName(identifier: any) {
      return identifier.image;
    }

    /*

    public headerClause(ctx: any) {
      const name = this.getIdentifierName(ctx.IdentifierToken[0]);
      const value = this.visit(ctx.valueClause[0]);
      return {
        name,
        value,
      };
    }

    public MemberCallNewExpressionExtension(ctx: any) {
      if (has(ctx, "BoxMemberExpression")) {
        return this.visit(ctx.BoxMemberExpression);
      }
      if (has(ctx, "DotMemberExpression")) {
        return this.visit(ctx.DotMemberExpression);
      }
      if (has(ctx, "Arguments")) {
        return this.visit(ctx.Arguments);
      }
    }

    public Arguments(ctx: any) {
      return `(${join(map(ctx.AssignmentExpression, this.visit.bind(this)))})`;
    }

    public DotMemberExpression(ctx: any) {
      if (has(ctx, "IdentifierToken")) {
        return `.${this.getIdentifierName(ctx.IdentifierToken[0])}`;
      }
    }

    */
  }
  return new Visitor();
}
