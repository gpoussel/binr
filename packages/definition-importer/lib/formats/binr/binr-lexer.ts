import { Lexer } from "chevrotain";
import { values } from "lodash";

import { tokens } from "./binr-tokens";

export class BinrLexer extends Lexer {
  constructor() {
    super(values(tokens), {
      ensureOptimizations: true,
    });
  }
}
