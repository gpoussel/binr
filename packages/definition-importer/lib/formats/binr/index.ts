export * from "./binr-definition-importer";
