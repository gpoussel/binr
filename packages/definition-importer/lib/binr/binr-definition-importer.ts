import { CstParser } from "chevrotain";

import { CStylePreprocessor } from "../common/cstyle-preprocessor";
import { Importer } from "../importer";
import { BinrLexer } from "./binr-lexer";
import { BinrParser } from "./binr-parser";
import { getVisitor } from "./binr-visitor";

export class BinrDefinitionImporter extends Importer {
  public getLexer() {
    return new BinrLexer();
  }

  public getParser() {
    return new BinrParser();
  }

  public getVisitor(parser: CstParser) {
    return getVisitor(parser);
  }

  public performPreprocessing(input: string) {
    const preprocessor = new CStylePreprocessor();
    return preprocessor.preprocess(input);
  }
}
