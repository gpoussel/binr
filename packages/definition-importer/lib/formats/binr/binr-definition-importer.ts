import { CstParser } from "chevrotain";

import { NoopPreprocessor, Preprocessor } from "../../common/preprocessors";
import { Importer } from "../importer";
import { BinrLexer } from "./binr-lexer";
import { BinrParser } from "./binr-parser";
import { getVisitor } from "./binr-visitor";

export class BinrDefinitionImporter extends Importer {
  public getLexer() {
    return new BinrLexer();
  }

  public getParser() {
    return new BinrParser();
  }

  public getVisitor(parser: CstParser) {
    return getVisitor(parser);
  }

  getPreprocessor(): Preprocessor {
    return new NoopPreprocessor();
  }
}
