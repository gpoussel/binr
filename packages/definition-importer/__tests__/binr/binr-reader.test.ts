import { Definition } from "@binr/ast";
import { each } from "lodash";

import { BinrDefinitionImporter } from "../../lib/binr/binr-definition-importer";

function createAndCallParser(input: string): () => Definition {
  return () => {
    const reader = new BinrDefinitionImporter();
    return reader.readInput(input);
  };
}

describe("BinrDefinitionImporter", () => {
  test("throws an error when lexing invalid input", () => {
    expect(createAndCallParser("👍")).toThrow(/lexing/);
  });

  test("throws an error when parsing invalid input", () => {
    each(["#bar", "struct {}", "123", "beep", "##"], (value: string) => {
      expect(createAndCallParser(value)).toThrow(/parsing/);
    });
  });

  test("throws an error on invalid content", () => {
    each(
      [
        "struct a {} struct a {}",
        "struct a {int a; int a;}",
        "",
        '#set "b" #set "b" struct a {}',
        "struct int {}",
        "struct a { b b; }",
        "struct c { int:0 a; }",
        "struct c { int:255 a; }",
        "struct c { string a; }",
      ],
      (value: string) => {
        expect(createAndCallParser(value)).toThrow(/validation/i);
      },
    );
  });

  test("accepts minimal input", () => {
    each(['#set "foo" struct a {}', "#set 32 struct b {}", "struct a {}"], (value: string) => {
      expect(createAndCallParser(value)()).toBeDefined();
    });
  });

  test("accepts field annotations", () => {
    each(["struct a { @ignore(false) int b; @ignore(true) int c; }"], (value: string) => {
      expect(createAndCallParser(value)()).toBeDefined();
    });
  });

  test("accepts bitmasks", () => {
    each(["bitmask foo extends int:16 { A = 10 } struct A {}"], (value: string) => {
      const result = createAndCallParser(value)();
      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  test("accepts enumerations", () => {
    each(["enum foo extends int:16 { A = 10 } struct A {}"], (value: string) => {
      const result = createAndCallParser(value)();
      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  test("accepts == and ||", () => {
    const result = createAndCallParser("struct a { int foo[1 == 2 || 3 == 4]; }")();
    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
  });

  test("rejects unsupported expression", () => {
    each(["a++", "a--", "++a", "--a"], (value: string) => {
      expect(createAndCallParser(`struct a { int foo[${value}]; }`)).toThrow(/supported/);
    });
  });
  test("rejects invalid expression", () => {
    each(["+", "())"], (value: string) => {
      expect(createAndCallParser(`struct a { int foo[${value}]; }`)).toThrow(/parsing/);
    });
  });
});
