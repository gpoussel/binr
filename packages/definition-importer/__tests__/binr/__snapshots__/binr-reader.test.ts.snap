// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinrDefinitionImporter accepts == and || 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [
          VariableDeclarationStatement {
            "_annotations": Array [],
            "_bitfield": undefined,
            "_modifiers": Array [],
            "_variableDeclarations": Array [
              VariableDeclaration {
                "_annotations": Array [],
                "_arraySelector": ExpressionArraySelector {
                  "_innerExpression": BinaryExpression {
                    "_left": BinaryExpression {
                      "_left": BinaryExpression {
                        "_left": NumberValue {
                          "_value": 1,
                        },
                        "_operator": 8,
                        "_right": NumberValue {
                          "_value": 2,
                        },
                      },
                      "_operator": 5,
                      "_right": NumberValue {
                        "_value": 3,
                      },
                    },
                    "_operator": 8,
                    "_right": NumberValue {
                      "_value": 4,
                    },
                  },
                },
                "_bitfield": undefined,
                "_initializationExpression": undefined,
                "_name": "foo",
                "_typeArguments": Array [],
              },
            ],
            "_variableType": NamedType {
              "_array": false,
              "_modifiers": Array [],
              "_name": "int",
            },
          },
        ],
      },
      "_name": "a",
      "_parameters": Array [],
    },
  ],
}
`;

exports[`BinrDefinitionImporter accepts bitmasks 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    BitmaskDeclarationStatement {
      "_annotations": Array [],
      "_baseType": RestrictedType {
        "_baseType": NamedType {
          "_array": false,
          "_modifiers": Array [],
          "_name": "int",
        },
        "_size": 16,
      },
      "_declarations": Array [
        BitmaskDeclarationElement {
          "_expression": 10,
          "_name": "A",
        },
      ],
      "_name": "foo",
    },
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [],
      },
      "_name": "A",
      "_parameters": Array [],
    },
  ],
}
`;

exports[`BinrDefinitionImporter accepts enumerations 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    EnumDeclarationStatement {
      "_annotations": Array [],
      "_baseType": RestrictedType {
        "_baseType": NamedType {
          "_array": false,
          "_modifiers": Array [],
          "_name": "int",
        },
        "_size": 16,
      },
      "_declarations": Array [
        EnumDeclarationElement {
          "_expression": 10,
          "_name": "A",
        },
      ],
      "_name": "foo",
    },
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [],
      },
      "_name": "A",
      "_parameters": Array [],
    },
  ],
}
`;

exports[`BinrDefinitionImporter accepts field annotations 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [
          VariableDeclarationStatement {
            "_annotations": Array [
              Annotation {
                "_key": "ignore",
                "_value": BooleanValue {
                  "_value": false,
                },
              },
            ],
            "_bitfield": undefined,
            "_modifiers": Array [],
            "_variableDeclarations": Array [
              VariableDeclaration {
                "_annotations": Array [],
                "_arraySelector": undefined,
                "_bitfield": undefined,
                "_initializationExpression": undefined,
                "_name": "b",
                "_typeArguments": Array [],
              },
            ],
            "_variableType": NamedType {
              "_array": false,
              "_modifiers": Array [],
              "_name": "int",
            },
          },
          VariableDeclarationStatement {
            "_annotations": Array [
              Annotation {
                "_key": "ignore",
                "_value": BooleanValue {
                  "_value": true,
                },
              },
            ],
            "_bitfield": undefined,
            "_modifiers": Array [],
            "_variableDeclarations": Array [
              VariableDeclaration {
                "_annotations": Array [],
                "_arraySelector": undefined,
                "_bitfield": undefined,
                "_initializationExpression": undefined,
                "_name": "c",
                "_typeArguments": Array [],
              },
            ],
            "_variableType": NamedType {
              "_array": false,
              "_modifiers": Array [],
              "_name": "int",
            },
          },
        ],
      },
      "_name": "a",
      "_parameters": Array [],
    },
  ],
}
`;

exports[`BinrDefinitionImporter accepts minimal input (#set "foo"
struct a {}) 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [],
      },
      "_name": "a",
      "_parameters": Array [],
    },
  ],
}
`;

exports[`BinrDefinitionImporter accepts minimal input (#set 32
struct b {}) 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [],
      },
      "_name": "b",
      "_parameters": Array [],
    },
  ],
}
`;

exports[`BinrDefinitionImporter accepts minimal input (struct a {}) 1`] = `
Definition {
  "_annotations": Array [],
  "_statements": Array [
    StructDeclarationStatement {
      "_annotations": Array [],
      "_body": BlockStatement {
        "_content": Array [],
      },
      "_name": "a",
      "_parameters": Array [],
    },
  ],
}
`;
